### DISCUSSION TITLE
Regarding extraction from `&PyAny` (or `PyObject`) to Rust object

### DISCUSSION THREAD

**Discussion by minsungkim-qraft** [PyO3/pyo3#3940] [Questions] [answered]
Hello. I wanted to extract the Rust object from Python function call.

### Impl

```rust
/// 'source is added because of below paragraph
impl<'source, I, O> SomeHiddenTrait<I, O> for SomeHiddenClass
where
    I: ToPyObject,
    O: PyClass + Clone,
{
    fn update(&mut self, data: &I) -> anyhow::Result<O> {
        Python::with_gil(|py| {
            let args = PyTuple::new(py, &[data]);
            let called = self.func.as_ref(py).call1(args)?;
            let extracted = called.extract::<O>()?;
            Ok(extracted)
        })
    }
}
```

If I change `O: PyClass + Clone` into `O: FromPyObject<'source> + Clone` (this was my first trial), then this does not work anymore. Rust compiler keeps complaining that the borrowed data must not move out of the closure, which confused me because `O` satisfies `Clone` anyway, so I imagined the cloned value is moving out of closure instead of borrowed value.

What is the reason why `O: FromPyObject<'source> + Clone` failed, and `O: PyClass + Clone` works?

**Comment by davidhewitt** [comment, marked as answer]
The problem is that in the example you give above, `'source` is a lifetime limited to inside the `with_gil` closure.

To avoid this, either pass `py: Python<'source>` as an argument, or use `O: for<'a> FromPyObject<'a> + Clone` as the bound, which will limit the function to types which don't capture the `FromPyObject` lifetime in their output.

**Reply by minsungkim-qraft** [reply]
The first option, `py: Python<'source>` made an error - compiler complains that `'source` must outlive `'static`.

The second option works good, I didn't know that generic `for` prefix was available on trait constraints. Guess I have to find some source from internet.

Thank you.
